### PR TITLE
Hash passwords with bcrypt instead of md5

### PR DESCRIPTION
**Pull request by KTS915**
## Description
New passwords are hashed using bcrypt instead of md5. Current passwords that are already hashed with md5 are instead hashed with bcrypt when the user next logs in.

## Motivation and context
See https://petitions.classicpress.net/posts/69/hash-passwords-with-bcrypt-instead-of-md5

## How has this been tested?
Tested on localhost installation running ClassicPress 1.0.0-alpha

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ x ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by senlin** [comment]
@KTS915 - Has this feature already reached the minimum votes threshold?

**Comment by ginsterbusch** [comment]
also, this seems to break code .. (at least according to Travis CI) ...

**Comment by senlin** [comment]
@ginsterbusch - yes there are 3 errors that all have to do with the unit tests.

**Comment by KTS915** [comment]
@senlin, I don't know. What is the threshold?

I thought I'd do it anyway while I had the time. Interesting that it's failed the Travis CI build. It passed that on my fork.

**Comment by senlin** [comment]
@KTS915, from the petitions website:

> Any suggestions that reach the Support Threshold (50 upvotes) will be put to a vote for the wider community.

Don't get me wrong, I voted for your idea, but I think it is important that we keep the process in place before submitting PRs left, right and center :)

**Comment by KTS915** [comment]
@senlin, I agree with you 100%. That bit of text was added after I last looked at the petitions site!

It gives me a chance anyway to see if I can work out why it fails the unit tests.

**Comment by senlin** [comment]
Most features will become a topic after v1 has shipped though, at least that has been my understanding.

**Comment by KTS915** [comment]
Is it best to close this PR now and then resubmit a new PR later, or would you prefer to keep this one open? If you'd prefer to keep this open, how do you want me to submit revisions?

**Comment by senlin** [comment]
Yes, I think it is better to close this and perhaps open an issue where you make a reference to the PR, so later on we will be able to find it. Then in the issue you can also reference the feature over on the petitions site, so once it has gathered enough votes, we will be able to find your initial PR again. Does that work for you?

**Comment by KTS915** [comment]
Yes, that seems best to me too! Thanks again!

File: src/wp-includes/pluggable.php [hunk 1]
```diff
@@ -2180,34 +2180,27 @@ function wp_hash($data, $scheme = 'auth') {
...
-
-	if ( empty($wp_hasher) ) {
-		require_once( ABSPATH . WPINC . '/class-phpass.php');
-		// By default, use the portable hash from phpass
-		$wp_hasher = new PasswordHash(8, true);
-	}
-
-	return $wp_hasher->HashPassword( trim( $password ) );
+function wp_hash_password( $password ) {
+	$options = apply_filters( 'wp_hash_password_options', [] );
```
Comment (nylen) [inline src/wp-includes/pluggable.php]: New filter needs inline documentation.  And since it's a new one, may as well change it to `cp_` too.

**Comment by KTS915** [comment]
@nylen,

> Caution: Using the PASSWORD_BCRYPT as the algorithm, will result in the password parameter being truncated to a maximum length of 72 characters.

Hmm, maybe that's the reason for one of the unit test failures. Not sure about the other two failures, though.

The code I used is based on code originally published by Roots at https://github.com/roots/wp-password-bcrypt and which I have used as an mu-plugin for two years on about half a dozen membership sites without any problems.

One issue I encountered was ensuring that the code always recognized an md5 hashed password when someone logged in with one (so that it would then be updated to bcrypt). There are WP plugins that purport to do the same job as the Roots code, but their method sometimes misses an md5 hashed password, whereas this code never does.

I also noticed that when the password hashing is changed from md5 to bcrypt, the user activation key in the `users` table in the database is deleted. Of course, that's exactly what the code says should happen, but I don't know if that key is needed for anything. It doesn't seem to be, because I have plenty of other entries in the `users` table without such a key.

**Comment by Mte90** [comment]
Another point this not break compatibility with old password already saved in md5?

**Comment by ginsterbusch** [comment]
> I also noticed that when the password hashing is changed from md5 to bcrypt, the user activation key in the `users` table in the database is deleted. Of course, that's exactly what the code says should happen, but I don't know if that key is needed for anything. It doesn't seem to be, because I have plenty of other entries in the `users` table without such a key.

Not that I know of. Maybe for logging purposes, eg. which user was created by registration or via the Admin UI, and which was created by different means? 

Different means would be: I've come up with a pretty easy method to quickly get access to a WP installation, eg. when the user has nio clue how to create an admin account, or lost his password AND access to his own mail address / changed it and the old one doesnt exist anymore. 

Thats done using a simple script, which is placed in the WP root. This script then bootstraps WP, bypases theme inclusion and creates a new admin user via `wp_create_user` (essentially I've been recreating what happens when you add a new user in the Admin UI). Optionally the newly created user data are sent to you via `mail()`, as the internal wp_mail is not reliable enough (sometimes it fails sending, although the standard `mail` function works flawlessly).

cu, w0lf.

**Comment by KTS915** [comment]
@Mte90 There is no problem with old passwords saved with an md5 hash. They are simply re-hashed using bcrypt when the user next logs in. I have tested that extensively and it works perfectly.

**Pull request closed without merging**